src/pymor/algorithms/eigs.py

```diff
@@ -0,0 +1,304 @@
+    E
+        The |Operator| which defines the generalized eigenvalue problem.
+    k
+        The number of eigenvalues and eigenvectors which are to be computed.
+    which
+        A string specifying which `k` eigenvalues and eigenvectors to compute:
+
+        - `'LM'`: select eigenvalues with largest absolute value
+        - `'SM'`: select eigenvalues with smallest absolute value
+        - `'LR'`: select eigenvalues with largest real part
+        - `'SR'`: select eigenvalues with smallest real part
+        - `'LI'`: select eigenvalues with largest imaginary part
+        - `'SI'`: select eigenvalues with smallest imaginary part
+    b
+        Initial vector for Arnoldi iteration. Default is a random vector.
+    l
+        The size of the Arnoldi factorization. Default is `min(n - 1, max(2*k + 1, 20))`.
+    maxiter
+        The maximum number of iterations.
+    tol
+        The relative error tolerance for the Ritz estimates.
+    seed
+        Random seed which is used for computing the initial vector for the Arnoldi
+        iteration.
+
+    Returns
+    -------
+    w
+        A 1D |NumPy array| which contains the computed eigenvalues.
+    v
+        A |VectorArray| which contains the computed eigenvectors.
+    """
+
+    n = A.source.dim
+
+    if l is None:
+        l = min(n - 1, max(2 * k + 1, 20))
+
+    assert isinstance(A, Operator) and A.linear
+    assert not A.parametric
+    assert A.source == A.range
+
+    if E is None:
+        E = IdentityOperator(A.source)
+    else:
+        assert isinstance(E, Operator) and E.linear
+        assert not E.parametric
+        assert E.source == E.range
+        assert E.source == A.source
+
+    if b is None:
+        b = A.source.random(seed=seed)
+    else:
+        assert b in A.source
+
+    assert k < n
+    assert l > k
+
+    V, H, f = _arnoldi(A, E, k, b)
+    k0 = k
+    i = 0
+
+    while True:
+        i += 1
+
+        V, H, f = _extend_arnoldi(A, E, V, H, f, l - k)
+
+        ew, ev = spla.eig(H)
+
+        # truncate small imaginary parts
+        ew.imag[np.abs(ew.imag) / np.abs(ew) < 1e-12] = 0
+
+        if which == 'LM':
+            idx = np.argsort(-np.abs(ew))
+        elif which == 'SM':
+            idx = np.argsort(np.abs(ew))
+        elif which == 'LR':
+            idx = np.argsort(-ew.real)
+        elif which == 'SR':
+            idx = np.argsort(ew.real)
+        elif which == 'LI':
+            idx = np.argsort(-np.abs(ew.imag))
+        elif which == 'SI':
+            idx = np.argsort(np.abs(ew.imag))
+
+        k = k0
+        ews = ew[idx]
+        evs = ev[:, idx]
+
+        rres = f.l2_norm()[0] * np.abs(evs[l - 1]) / np.abs(ews)
+
+        # increase k by one in order to keep complex conjugate pairs together
+        if ews[k - 1].imag != 0 and ews[k - 1].imag + ews[k].imag < 1e-12:
+            k += 1
+
+        if np.all(rres[:k] <= tol) or i >= maxiter:
+            break
```

Comment:
It would be nice to report the residual (I guess `rres[:k].max()`?) and the iteration number via `logger.info` before this check.